File: python/gatherlink/security/keys.py
"""Node identity and signing helpers owned by the Python control plane."""

from __future__ import annotations

from dataclasses import dataclass
from hashlib import sha256

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.serialization import Encoding, NoEncryption, PrivateFormat, PublicFormat

NODE_ID_DOMAIN = b"gatherlink node v1"


@dataclass(frozen=True)
class NodeIdentity:
    """A Gatherlink node identity with signing and transport static keys."""

    ed25519_private: bytes
    ed25519_public: bytes
    x25519_private: bytes
    x25519_public: bytes
    node_id: bytes

    @classmethod
    def generate(cls) -> NodeIdentity:
        """Generate a new Ed25519 identity and X25519 transport static key."""
        ed_private = Ed25519PrivateKey.generate()
        x_private = X25519PrivateKey.generate()
        ed_public = ed_private.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
        x_public = x_private.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
        return cls(
            ed25519_private=ed_private.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption()),
            ed25519_public=ed_public,
            x25519_private=x_private.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption()),
            x25519_public=x_public,
            node_id=node_id_from_ed25519_public(ed_public),
        )

    @classmethod
    def from_private_keys(cls, ed25519_private: bytes, x25519_private: bytes) -> NodeIdentity:
        """Load an identity from raw private-key bytes."""
        ed_private = Ed25519PrivateKey.from_private_bytes(ed25519_private)
        x_private = X25519PrivateKey.from_private_bytes(x25519_private)
        ed_public = ed_private.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
        x_public = x_private.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
        return cls(
            ed25519_private=ed25519_private,
            ed25519_public=ed_public,
            x25519_private=x25519_private,
            x25519_public=x_public,
            node_id=node_id_from_ed25519_public(ed_public),
        )

    def sign(self, domain: bytes, canonical_body: bytes) -> bytes:
        """Sign a domain-separated canonical document body."""
        return sign_document(self.ed25519_private, domain, canonical_body)


def node_id_from_ed25519_public(public_key: bytes) -> bytes:
    """Return the stable Gatherlink node id for an Ed25519 public key."""
    if len(public_key) != 32:
        raise ValueError("Ed25519 public key must be 32 bytes")
    return sha256(NODE_ID_DOMAIN + public_key).digest()


def sign_document(private_key: bytes, domain: bytes, canonical_body: bytes) -> bytes:
    """Sign domain-separated canonical document bytes with Ed25519."""
    return Ed25519PrivateKey.from_private_bytes(private_key).sign(domain + canonical_body)


def verify_document(public_key: bytes, domain: bytes, canonical_body: bytes, signature: bytes) -> None:
    """Verify a domain-separated Ed25519 document signature."""
    Ed25519PublicKey.from_public_bytes(public_key).verify(signature, domain + canonical_body)


def x25519_shared_secret(private_key: bytes, peer_public_key: bytes) -> bytes:
    """Derive a raw X25519 shared secret."""
    private = X25519PrivateKey.from_private_bytes(private_key)
    public = X25519PublicKey.from_public_bytes(peer_public_key)
    return private.exchange(public)

File: crates/crypto/src/keys.rs
//! Key and identity primitives for the Python-controlled security layer.

use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use rand_core::OsRng;
use sha2::{Digest, Sha256};
use x25519_dalek::{PublicKey as X25519PublicKey, StaticSecret};

use crate::errors::CryptoError;

/// Ed25519 public key length.
pub const ED25519_PUBLIC_KEY_LEN: usize = 32;
/// Ed25519 private seed length.
pub const ED25519_PRIVATE_KEY_LEN: usize = 32;
/// Ed25519 signature length.
pub const ED25519_SIGNATURE_LEN: usize = 64;
/// X25519 public/private key length.
pub const X25519_KEY_LEN: usize = 32;
/// Stable node id length.
pub const NODE_ID_LEN: usize = 32;
/// Domain separator for Gatherlink node ids.
pub const NODE_ID_DOMAIN: &[u8] = b"gatherlink node v1";

/// Generate an Ed25519 signing seed and public key.
#[must_use]
pub fn generate_ed25519_keypair() -> ([u8; ED25519_PRIVATE_KEY_LEN], [u8; ED25519_PUBLIC_KEY_LEN]) {
    let signing_key = SigningKey::generate(&mut OsRng);
    (signing_key.to_bytes(), signing_key.verifying_key().to_bytes())
}

/// Generate an X25519 static private/public keypair.
#[must_use]
pub fn generate_x25519_keypair() -> ([u8; X25519_KEY_LEN], [u8; X25519_KEY_LEN]) {
    let secret = StaticSecret::random_from_rng(OsRng);
    let public = X25519PublicKey::from(&secret);
    (secret.to_bytes(), public.to_bytes())
}

/// Compute the stable Gatherlink node id from an Ed25519 public key.
#[must_use]
pub fn node_id_from_ed25519_public(public_key: &[u8; ED25519_PUBLIC_KEY_LEN]) -> [u8; NODE_ID_LEN] {
    let mut hasher = Sha256::new();
    hasher.update(NODE_ID_DOMAIN);
    hasher.update(public_key);
    hasher.finalize().into()
}

/// Sign domain-separated document bytes with an Ed25519 seed.
pub fn sign_document(
    private_seed: &[u8; ED25519_PRIVATE_KEY_LEN],
    domain: &[u8],
    canonical_body: &[u8],
) -> [u8; ED25519_SIGNATURE_LEN] {
    let signing_key = SigningKey::from_bytes(private_seed);
    let mut message = Vec::with_capacity(domain.len() + canonical_body.len());
    message.extend_from_slice(domain);
    message.extend_from_slice(canonical_body);
    signing_key.sign(&message).to_bytes()
}

/// Verify a domain-separated Ed25519 document signature.
pub fn verify_document(
    public_key: &[u8; ED25519_PUBLIC_KEY_LEN],
    domain: &[u8],
    canonical_body: &[u8],
    signature: &[u8; ED25519_SIGNATURE_LEN],
) -> Result<(), CryptoError> {
    let verifying_key = VerifyingKey::from_bytes(public_key).map_err(|_| CryptoError::InvalidInput)?;
    let signature = Signature::from_bytes(signature);
    let mut message = Vec::with_capacity(domain.len() + canonical_body.len());
    message.extend_from_slice(domain);
    message.extend_from_slice(canonical_body);
    verifying_key
        .verify(&message, &signature)
        .map_err(|_| CryptoError::SilentDrop)
}

/// Derive a shared X25519 secret.
#[must_use]
pub fn x25519_shared_secret(
    private_key: &[u8; X25519_KEY_LEN],
    peer_public_key: &[u8; X25519_KEY_LEN],
) -> [u8; X25519_KEY_LEN] {
    let secret = StaticSecret::from(*private_key);
    let peer = X25519PublicKey::from(*peer_public_key);
    secret.diffie_hellman(&peer).to_bytes()
}
